pkg/imgutil/dockerconfigresolver/credentialsstore_test.go

```diff
@@ -0,0 +1,384 @@
+		t.Fatal(err)
+	}
+	err = os.Chmod(tmpDir, mode)
+	if err != nil {
+		t.Fatal(err)
+	}
+	return tmpDir
+}
+
+func TestBrokenCredentialsStore(t *testing.T) {
+	if runtime.GOOS == "freebsd" {
+		// It is unclear why these tests are failing on FreeBSD, and if it is a problem with Vagrant or differences
+		// with FreeBSD
+		// Anyhow, this test is about extreme cases & conditions (filesystem errors wrt credentials loading).
+		t.Skip("skipping broken credential store tests for freebsd")
+	}
+
+	testCases := []struct {
+		description string
+		setup       func() string
+		errorNew    error
+		errorRead   error
+		errorWrite  error
+	}{
+		{
+			description: "Pointing DOCKER_CONFIG at a non-existent directory inside an unreadable directory will prevent instantiation",
+			setup: func() string {
+				tmpDir := createTempDir(t, 0000)
+				return filepath.Join(tmpDir, "doesnotexistcantcreate")
+			},
+			errorNew: ErrUnableToInstantiate,
+		},
+		{
+			description: "Pointing DOCKER_CONFIG at a non-existent directory inside a read-only directory will prevent saving credentials",
+			setup: func() string {
+				tmpDir := createTempDir(t, 0500)
+				return filepath.Join(tmpDir, "doesnotexistcantcreate")
+			},
+			errorWrite: ErrUnableToStore,
+		},
+		{
+			description: "Pointing DOCKER_CONFIG at an unreadable directory will prevent instantiation",
+			setup: func() string {
+				return createTempDir(t, 0000)
+			},
+			errorNew: ErrUnableToInstantiate,
+		},
+		{
+			description: "Pointing DOCKER_CONFIG at a read-only directory will prevent saving credentials",
+			setup: func() string {
+				return createTempDir(t, 0500)
+			},
+			errorWrite: ErrUnableToStore,
+		},
+		{
+			description: "Pointing DOCKER_CONFIG at a directory containing am unparsable `config.json` will prevent instantiation",
+			setup: func() string {
+				tmpDir := createTempDir(t, 0700)
+				err := os.WriteFile(filepath.Join(tmpDir, "config.json"), []byte("porked"), 0600)
+				if err != nil {
+					t.Fatal(err)
+				}
+				return tmpDir
+			},
+			errorNew: ErrUnableToInstantiate,
+		},
+		{
+			description: "Pointing DOCKER_CONFIG at a file instead of a directory will prevent instantiation",
+			setup: func() string {
+				tmpDir := createTempDir(t, 0700)
+				fd, err := os.OpenFile(filepath.Join(tmpDir, "isafile"), os.O_CREATE, 0600)
+				if err != nil {
+					t.Fatal(err)
+				}
+				err = fd.Close()
+				if err != nil {
+					t.Fatal(err)
+				}
+				return filepath.Join(tmpDir, "isafile")
+			},
+			errorNew: ErrUnableToInstantiate,
+		},
+		{
+			description: "Pointing DOCKER_CONFIG at a directory containing a `config.json` directory will prevent instantiation",
+			setup: func() string {
+				tmpDir := createTempDir(t, 0700)
+				err := os.Mkdir(filepath.Join(tmpDir, "config.json"), 0600)
+				if err != nil {
+					t.Fatal(err)
+				}
+				return tmpDir
+			},
+			errorNew: ErrUnableToInstantiate,
+		},
+		// FIXME: this test works in Lima, but does not work on github action
```

Comment:
Mystery for the ages.

Reply:
What is the error on GHA?

Reply:
So, in the test conditions, which are: `DOCKER_CONFIG=/directoryxyz`, and directoryxyz contains a config.json file that is a dangling symlink.


Here is what actually happen on LIMA:

```
unable to store credentials
                rename /tmp/TestBrokenCredentialsStore705218110/008/docker-config2430087685/config.json2911426040 doesnotexist: invalid cross-device link
```

On GHA, things just work with no error.

The test assumes an error, so, locally the test "succeeds", but on GHA it "fails".

My guess to what is happening is that the docker lib finds the broken `config.json` and moves it over to create a new file. 
On GHA, the command succeeds (which is probably the expected behavior after all...), but then not on LIMA...

@AkihiroSuda do you understand why?
`invalid cross-device link` suggests this is about hard link, but this here is a symlink...

Reply:
(I am using virtio FWIW)

Reply:
-> https://github.com/docker/cli/blob/master/cli/config/configfile/file.go#L170-L178 <-

Reply:
Could you add the error to the code comments?

Reply:
Update: testing again with latest Lima (using your fork, actually) on Ubuntu 24.04, and it just works...

Unclear to me what that was - maybe older ubuntu, or maybe something in Lima fixed this.

Will enable the test then.